maven_iuvs/echelle.py

```diff
@@ -1415,297 +1456,342 @@ def convert_l1a_to_l1c(light_fits, dark_fits, light_l1a_path, savepath, calibrat
         else:
             pass 
 
-        fitting_results = fit_H_and_D(initial_guess, wavelengths, spec, light_fits, theCLSF, unc=unc, fit_IPH=fit_IPH,
-                                      solver=solv, fitter=fitpackage, approach=approach, livepts=livepts, hush_warning=hush_warning) 
-        if return_each_line_fit:
-            fit_params_list, I_fit, fit_1sigma, H_fit, D_fit, IPH_fit = fitting_results
+        if BU_bg is not None:
+            BU_bg_i = BU_bg[i, :]
         else:
-            fit_params_list, I_fit, fit_1sigma, *_ = fitting_results
-
-        # Convert fit_params to a dictionary so it's easier to use.
-        fit_params = {"area_H": fit_params_list[0], "area_D": fit_params_list[1], 
-                       "lambdac_H": fit_params_list[2], "lambdac_D": fit_params_list[2]-D_offset, 
-                       "M":fit_params_list[3], "B": fit_params_list[4]}
-        fit_unc = {"uncert_H": fit_1sigma[0], "uncert_D": fit_1sigma[1], "uncert_lambdac_H": fit_1sigma[2], 
-                   "uncert_M": fit_1sigma[3], "uncert_B": fit_1sigma[4]}
-
-        if fit_IPH:
-            fit_params["area_IPH"] = fit_params_list[-2]
-            fit_params["lambdac_IPH"] = fit_params_list[-1]         
-            fit_unc["uncert_IPH"] = fit_1sigma[-2]
-            fit_unc["uncert_lambdac_IPH"] = fit_1sigma[-1]  
+            BU_bg_i = None
 
-        
-        # Create a convenient dictionary which can be used with a plotting routine
-        if np.isnan(fit_params_list).all():
-            fit_params_for_printing = fit_params
-        else:
-            fit_params_for_printing = {'area_H': round(fit_params["area_H"]), 'area_D': round(fit_params["area_D"]), 
-                                        'lambdac_H': round(fit_params["lambdac_H"], 3), 'lambdac_D': round(fit_params["lambdac_H"]-D_offset, 3),
-                                        "M": round(fit_params["M"]), "B": round(fit_params["B"])}
+        result_vec = fit_H_and_D(initial_guess, wavelengths, spectrum[i, :], light_fits, theCLSF, unc=data_unc[i, :], BU_bg=BU_bg_i, fit_IPH=fit_IPH, **kwargs)
             
-        if fit_IPH:
-            fit_params_for_printing["area_IPH"] = fit_params["area_IPH"]
-            fit_params_for_printing["lambdac_IPH"] = fit_params["lambdac_IPH"]
-        
-        # Construct a background array from the fit which can then be converted like the spectrum
-        bg_fit = background(wavelengths, fit_params['M'], fit_params['lambdac_H'], fit_params['B'])
+        if return_each_line_fit:
+            fit_params, I_fit, fit_1sigma, H_fit, D_fit, IPH_fit = result_vec
+            H_fit_array[i, :] = H_fit
+            D_fit_array[i, :] = D_fit
+            IPH_fit_array[i, :] = IPH_fit
+        else:
+            fit_params, I_fit, fit_1sigma, *_ = result_vec
+
+        I_fit_array[i, :] = I_fit
+        fit_unc_list.append(fit_1sigma)
+        fit_params_list.append(fit_params)
+
+    # Error control
+    if np.isnan(fit_params).all():
```

Comment:
a more strict criterion would be to fail if any of the integration fits return nan--- did you consider adopting this condition?

Reply:
I think this was written this way because it used to be that the routines would return nan for the D fit sometimes (but still get a valid H fit) but now it properly returns a float. Now a single nan is probably appropriate to fail the fit

Reply:
Fixed by https://github.com/lasp/maven_iuvs/pull/70/commits/6cf9de7a0afc25ea0394948c9912069b02dd32a7